# Напишите заголовок проекта здесь

- Автор: Родина Дарья
- Дата:

### Цели и задачи проекта

<font color='#777778'>Перечислите цель проекта и задачи, которые вы решаете. Для этого можно использовать описание проекта, но полезнее будет сформулировать основную цель проекта самостоятельно.</font>

Цель - провести полную предобработку данных 

Задачи:
- познакомиться с данными, 
- проверить их корректность 
- провести предобработку 
- получить необходимый срез данных

### Описание данных

<font color='#777778'>Здесь приведите описание данных.</font>

Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

 - Name — название игры.

- Platform — название платформы.

- Year of Release — год выпуска игры.

- Genre — жанр игры.

- NA sales — продажи в Северной Америке (в миллионах проданных копий).

- EU sales — продажи в Европе (в миллионах проданных копий).

- JP sales — продажи в Японии (в миллионах проданных копий).

- Other sales — продажи в других странах (в миллионах проданных копий).

- Critic Score — оценка критиков (от 0 до 100).

- User Score — оценка пользователей (от 0 до 10).

- Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

<font color='#777778'>Перечислите основные шаги проекта или напишите оглавление. Используйте описание проекта, чтобы зафиксировать основные шаги.</font>

---

1. Загрузка данных и знакомство с ними
2. Проверка ошибок в данных и их предобработка 
- 2.1 Названия, или метки, столбцов датафрейма 
- 2.2 Типы данных 
- 2.3 Наличие пропусков в данных 
- 2.4 Явные и неявные дубликаты в данных 
3. Фильтрация в данных 
4. Категоризация данных 
5. Итоговый вывод 

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
import pandas as pd 

In [2]:
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
df.head(10)

Name Platform  Year of Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

   NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0     41.36    28.96     3.77         8.45          76.0          8      E  
1     29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2     15.68    12.76     3.79         3.29          82.0        8.3      E  
3     15.61    10.93     3.28         2.95          80.0          8      E  
4     11.27     8.89    10.22         1.00           NaN        NaN    NaN  
5     23.20     2.26     4.22         0.58           NaN        NaN    NaN  
6     11.28     9.14      6.5         2.88          89.0        8.5      E  
7     13.96     9.18     2.93         2.84          58.0        6.6      E  
8     14.44     6.94      4.7         2.24          87.0        8.4      E  
9     26.93     0.63     0.28         0.47           NaN        NaN    NaN

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [5]:
# количество строк с самого начала 
total_row_count = df.shape[0]
total_row_count

16956

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

Данные предоставлены в количестве 16 956 строк и в объёме 1.4 МВ. Можно заметить пропущенные значения в столбцах Name, Year of Release, Genre, Critic Score, User Score и Rating. Всего данные предоставлены в типах float - 4 столбца и object - 7 столбцов. Не у всех столбцов указан верный тип данных. 

Все названия столбцов отражают содержимое данных и прописаны в удобном для работы виде.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [6]:
# выводим названия всех столбцов 
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [7]:
# приводим к нижнему регистру 
df.columns = df.columns.str.lower().str.replace(" ", "_")

In [8]:
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [9]:
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors = 'coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors = 'coerce')
df['user_score'] = pd.to_numeric(df['user_score'], errors = 'coerce')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [11]:
df.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [12]:
df.isna().sum()/len(df)*100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


В столбцах 'name' и 'genre' доля пропусков составляет менее 1%. Вероятнее всего, эти пропуски допущены из-за человеческого фактора, то есть их можно отнести к категории MСAR - полностью случайные пропуски. В таком случае, данные пропущенные значения можно удалить, так как на итоговый анализ это не повлияет.

В столбце 'year_of_release' доля пропусков составляет чуть более 1%. Такие пропуски можно также заменить на "0", чтобы их не удалять. 

В столбце 'critic_score' доля пропусков составляет более 50%, что является очень высоким и нехорошим показателем. Таким образом, данные пропуски нужно заменить на некоторое нереалистичное значение "-1". 

В столбце 'user_score' количество пропусков достигает 9 тысяч, а доля составляет 54%. Данные пропущенные зачения можно отнести к категории MNAR, то есть пропуски не случайны и мы их не можем объянсить другими значениями. Такие пропуски следует оставить без изменений. 
В столбце rating доля пропусков составляет 40%, что также является достаточно большим процентом. Таким образом данные пропуски следует заменить на пустые строки. 

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [13]:
# удаляем пропуски в столбцах 'name' и 'genre'
df = df.dropna(subset=['name', 'genre'])

In [14]:
#заполняем пропуски столбца 'eu_sales' средним значением в зависимоcти от названия платформы и года выхода игры
def mean_group_sales(row):
    if pd.isna(row['eu_sales']):
        group = df[(df['platform'] == row['platform'])&
                  (df['year_of_release'] == row['year_of_release'])]
        return group['eu_sales'].mean()
    else:
        return row['eu_sales']
df['eu_sales'] = df.apply(mean_group_sales, axis = 1)

In [15]:
# заполняем пропуски столбца 'jp_sales' средним значением в зависимоcти от названия платформы и года выхода игры
def mean_group_sales(row):
    if pd.isna(row['jp_sales']):
        group = df[(df['platform'] == row['platform'])&
                  (df['year_of_release'] == row['year_of_release'])]
        return group['jp_sales'].mean()
    else:
        return row['jp_sales']
df['jp_sales'] = df.apply(mean_group_sales, axis = 1)

In [16]:
# заполняем пропуски столбца year_of_release нулевыми значениями
df['year_of_release'] = df['year_of_release'].fillna(0)

In [17]:
# заполняем пропущенные значения столбца critic_score на среднее значение
df['critic_score'] = df['critic_score'].fillna(-1)

In [18]:
# заполняем пропуски столбца user_score медианой
df['user_score'] = df['user_score'].fillna(-1)

In [19]:
# заполняем пропуски столбца rating пустыми строками
df['rating'] = df['rating'].fillna('')

In [20]:
# проверяем количество пропусков
df.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

In [21]:
df['year_of_release'] = df['year_of_release'].astype('int')

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

- Выводим уникальные значения столбцов platform, year_of_release, genre, name, rating

In [22]:
unique_platform = df['platform'].unique()
unique_platform

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [23]:
unique_year_of_release = df['year_of_release'].unique()
unique_year_of_release

array([2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010,
       2013, 2004, 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014,
       1992, 1997, 1993, 1994, 1982, 2016, 2003, 1986, 2000,    0, 1995,
       1991, 1981, 1987, 1980, 1983])

In [24]:
unique_genre = df['genre'].unique()
unique_genre

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

In [25]:
unique_name = df['name'].unique()
unique_name

array(['Wii Sports', 'Super Mario Bros.', 'Mario Kart Wii', ...,
       'Woody Woodpecker in Crazy Castle 5', 'LMA Manager 2007',
       'Haitaka no Psychedelica'], dtype=object)

In [26]:
unique_rating = df['rating'].unique()

- В столбце 'genre' встречаются неявные дубликаты, которые записаны в разных регистрах.

- Приведем все к единому регистру, тем самым устраним неявные дубликаты в столбце 'genre'

In [27]:
df['genre']= df['genre'].str.lower()

In [28]:
df['name'] = df['name'].str.lower()

In [29]:
df['platform'] = df['platform'].str.upper()

In [30]:
df['rating'] = df['rating'].str.upper()

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [31]:
# находим количество явных дубликатов
duplicate_rows = df.duplicated().sum()
duplicate_rows

241

In [32]:
# фиксируем количество строк до удаления явных дубликатов
initial_row_count = df.shape[0]
initial_row_count

16954

In [33]:
# удаляем явные дубликаты
df = df.drop_duplicates()

In [34]:
# фиксируем количество строк после удаления дубликатов
final_row_count = df.shape[0]
final_row_count

16713

- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

С неявными дубликатами была проведена работа по проведению данных к единому регистру, так как в одном столбце были найдены значения, записанные под разными регистрами. Таким образом, неявные дубликаты были исключены из датафрейма. 

Количество явных дубликатов - 241. Данные дубликаты были удалены. 

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [35]:
# абсолютное значение удаленных строк 
cleaned_row_count_absolutely = total_row_count - final_row_count
cleaned_row_count_absolutely

243

In [36]:
# относительное значение удаленных строк 
cleened_row_count = round((final_row_count / total_row_count * 100), 2)
cleened_row_count

98.57

- После проведения предобработки данных напишите общий промежуточный вывод.

На текущем этапе была проведена тщательная работа с пропусками и дубликатами, что является важной частью предобработки данных. 
Были применены различные методы, включая приведение данных к единому регистру, удаление некоторых строк, а также заполнение недостающих данных с помощью расчетных значений и другое. В результате данные в датафрейме были очищены и теперь готовы для дальнейшего анализа.  

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [37]:
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)]
df_actual.head()

name platform  year_of_release     genre  na_sales  \
0             wii sports      WII             2006    sports     41.36   
2         mario kart wii      WII             2008    racing     15.68   
3      wii sports resort      WII             2009    sports     15.61   
6  new super mario bros.       DS             2006  platform     11.28   
7               wii play      WII             2006      misc     13.96   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45          76.0         8.0      E  
2     12.76      3.79         3.29          82.0         8.3      E  
3     10.93      3.28         2.95          80.0         8.0      E  
6      9.14      6.50         2.88          89.0         8.5      E  
7      9.18      2.93         2.84          58.0         6.6      E

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [38]:
def categorize_user_score(row):
    if row['user_score'] >= 8:
        return "высокая оценка"
    elif row['user_score'] >= 3:
        return "средняя оценка"
    elif row['user_score'] >= 0:
        return "низкая оценка"
    else:
        return "нет оценки"
df_actual = df_actual.copy()
df_actual['user_category'] =  df_actual.apply(categorize_user_score, axis =1)   

In [39]:
df_actual

name platform  \
0                                            wii sports      WII   
2                                        mario kart wii      WII   
3                                     wii sports resort      WII   
6                                 new super mario bros.       DS   
7                                              wii play      WII   
...                                                 ...      ...   
16947                     men in black ii: alien escape       GC   
16949                woody woodpecker in crazy castle 5      GBA   
16950  score international baja 1000: the official game      PS2   
16952                                  lma manager 2007     X360   
16954                                  spirits & spells      GBA   

       year_of_release     genre  na_sales  eu_sales  jp_sales  other_sales  \
0                 2006    sports     41.36     28.96      3.77         8.45   
2                 2008    racing     15.68     12.76      3.79         3.29   
3                 2009    sports     15.61     10.93      3.28         2.95   
6                 2006  platform     11.28      9.14      6.50         2.88   
7                 2006      misc     13.96      9.18      2.93         2.84   
...                ...       ...       ...       ...       ...          ...   
16947             2003   shooter      0.01      0.00      0.00         0.00   
16949             2002  platform      0.01      0.00      0.00         0.00   
16950             2008    racing      0.00      0.00      0.00         0.00   
16952             2006    sports      0.00      0.01      0.00         0.00   
16954             2003  platform      0.01      0.00      0.00         0.00   

       critic_score  user_score rating   user_category  
0              76.0         8.0      E  высокая оценка  
2              82.0         8.3      E  высокая оценка  
3              80.0         8.0      E  высокая оценка  
6              89.0         8.5      E  высокая оценка  
7              58.0         6.6      E  средняя оценка  
...             ...         ...    ...             ...  
16947          -1.0        -1.0      T      нет оценки  
16949          -1.0        -1.0             нет оценки  
16950          -1.0        -1.0             нет оценки  
16952          -1.0        -1.0             нет оценки  
16954          -1.0        -1.0             нет оценки  

[12781 rows x 12 columns]

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [40]:
def categorize_critic_score(row):
    if row['critic_score'] >= 80:
        return "высокая оценка"
    elif row['user_score'] >= 30:
        return "средняя оценка"
    elif row['user_score'] >= 0:
        return "низкая оценка"
    else:
        return "нет оценки"
df_actual = df_actual.copy()
df_actual['critic_category'] =  df_actual.apply(categorize_user_score, axis =1)   

In [41]:
df_actual

name platform  \
0                                            wii sports      WII   
2                                        mario kart wii      WII   
3                                     wii sports resort      WII   
6                                 new super mario bros.       DS   
7                                              wii play      WII   
...                                                 ...      ...   
16947                     men in black ii: alien escape       GC   
16949                woody woodpecker in crazy castle 5      GBA   
16950  score international baja 1000: the official game      PS2   
16952                                  lma manager 2007     X360   
16954                                  spirits & spells      GBA   

       year_of_release     genre  na_sales  eu_sales  jp_sales  other_sales  \
0                 2006    sports     41.36     28.96      3.77         8.45   
2                 2008    racing     15.68     12.76      3.79         3.29   
3                 2009    sports     15.61     10.93      3.28         2.95   
6                 2006  platform     11.28      9.14      6.50         2.88   
7                 2006      misc     13.96      9.18      2.93         2.84   
...                ...       ...       ...       ...       ...          ...   
16947             2003   shooter      0.01      0.00      0.00         0.00   
16949             2002  platform      0.01      0.00      0.00         0.00   
16950             2008    racing      0.00      0.00      0.00         0.00   
16952             2006    sports      0.00      0.01      0.00         0.00   
16954             2003  platform      0.01      0.00      0.00         0.00   

       critic_score  user_score rating   user_category critic_category  
0              76.0         8.0      E  высокая оценка  высокая оценка  
2              82.0         8.3      E  высокая оценка  высокая оценка  
3              80.0         8.0      E  высокая оценка  высокая оценка  
6              89.0         8.5      E  высокая оценка  высокая оценка  
7              58.0         6.6      E  средняя оценка  средняя оценка  
...             ...         ...    ...             ...             ...  
16947          -1.0        -1.0      T      нет оценки      нет оценки  
16949          -1.0        -1.0             нет оценки      нет оценки  
16950          -1.0        -1.0             нет оценки      нет оценки  
16952          -1.0        -1.0             нет оценки      нет оценки  
16954          -1.0        -1.0             нет оценки      нет оценки  

[12781 rows x 13 columns]

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [42]:
group_data = df_actual.groupby(['user_category', 'critic_category'], as_index = False)['name'].count()
group_data

user_category critic_category  name
0  высокая оценка  высокая оценка  2286
1      нет оценки      нет оценки  6298
2   низкая оценка   низкая оценка   116
3  средняя оценка  средняя оценка  4081

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [43]:
#группируем платформы 
group_platform = df_actual.groupby('platform', as_index = False)['name'].count()

In [44]:
#сортируем плафтормы по количеству выпущенных игр
sorted_platforms = group_platform.sort_values(by ='name', ascending = False).head(7)
sorted_platforms

platform  name
9       PS2  2127
2        DS  2120
14      WII  1275
12      PSP  1180
17     X360  1121
10      PS3  1087
4       GBA   811

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали.

Данные были тщательно отфильтрованы, что позволило выделить историю продаж игр в начале 21 века. 
Также была выполнена категоризация данных по оценкам пользователей и критиков. Это было сделано для того, чтобы информация лучше воспринималась заказчиком, а также соответствовала его требованиям. 

Кроме того, была проделана фильтрация данных по платформам, благодаря чему удалось выявить топ-7 платформ по количеству выпущенных игр. 